remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/mapping/typebridge/ProtobufMessageBinder.java

```diff
@@ -0,0 +1,92 @@
+package org.infinispan.query.remote.impl.mapping.typebridge;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Stack;
+
+import org.hibernate.search.engine.backend.document.IndexFieldReference;
+import org.hibernate.search.engine.backend.document.IndexObjectFieldReference;
+import org.hibernate.search.engine.backend.document.model.dsl.IndexSchemaElement;
+import org.hibernate.search.engine.backend.document.model.dsl.IndexSchemaObjectField;
+import org.hibernate.search.mapper.pojo.bridge.binding.TypeBindingContext;
+import org.hibernate.search.mapper.pojo.bridge.mapping.programmatic.TypeBinder;
+import org.infinispan.protostream.descriptors.Descriptor;
+import org.infinispan.query.remote.impl.mapping.reference.GlobalReferenceHolder;
+import org.infinispan.query.remote.impl.mapping.reference.IndexReferenceHolder;
+import org.infinispan.query.remote.impl.mapping.reference.MessageReferenceProvider;
+
+public class ProtobufMessageBinder implements TypeBinder {
+
+   private final GlobalReferenceHolder globalReferenceHolder;
+   private final String rootMessageName;
+
+   public ProtobufMessageBinder(GlobalReferenceHolder globalReferenceHolder, String rootMessageName) {
+      this.globalReferenceHolder = globalReferenceHolder;
+      this.rootMessageName = rootMessageName;
+   }
+
+   @Override
+   public void bind(TypeBindingContext context) {
+      context.getDependencies().useRootOnly();
+      IndexReferenceHolder indexReferenceProvider = createIndexReferenceProvider(context, 5);
```

Comment:
Important: this `5` here means we will only ever handle at most 5 level of embedding. In order to improve on this, we will need to add a proper `@IndexedEmbedded` annotation to the annotations supported by protobuf, with a `maxDepth` attribute.

We could go further and support an `includePaths` attribute as we do in Hibernate Search, but that will require some more work as the code below can only handle a `maxDepth`.

Reply:
Right, I think we should create a follow up issue for that. Meanwhile I extracted the magic number as a constant and I put a comment on top of it.